### PR TITLE
Move reason cache and api pod status convert function to kubelet level

### PR DESCRIPTION
**Pull request by Random-Liu**
This PR is based on #18172, and is part of #18302. The last 5 commits are new.

The following things are done in this PR:
- Let `SyncPod()` return `PodSyncResult`.
- Add reason cache in `kubelet/reason_cache.go`
- Add test for new reason cache and convert function.
- Add new reason cache into kubelet.
- Move api pod status convert function to kubelet level.

@yujuhong @yifan-gu

### CODE REVIEW

**Comment by Random-Liu** [comment]
I won't start using new convert function in this PR, so as to make it independent with the pod status cache. I'll send another PR for the following things which rely on the pod status.

**Comment by Random-Liu** [comment]
@k8s-bot test this please.

**Comment by Random-Liu** [comment]
@yujuhong @yifan-gu Ready for the first review. :)

**Comment by Random-Liu** [comment]
@k8s-bot e2e test this please.

**Comment by Random-Liu** [comment]
@yujuhong Addressed comments and squashed. But it's so sad that the comments are missing even after I use the "--force-with-lease" when pushing.

**Comment by Random-Liu** [comment]
@k8s-bot e2e test this please.

**Comment by Random-Liu** [comment]
This PR broke one e2e test, I'll look into it.

**Comment by Random-Liu** [comment]
@k8s-bot e2e test this please.

**Comment by Random-Liu** [comment]
@yujuhong The bug should have been fixed by the new commit.

This PR broke e2e test "should have their auto-restart back-off timer reset on image update" at first.

After some debuging, I found that we should call `reasonCache.Update()` whenever there is error or not. Because even when there is no error, we should "update" the reason cache to clean up the old error reason.

If we don't do that, once a container is failed to be started, its error reason will never be cleared, and will cause `convertStatusToAPIStatus()` always return `Waiting` state after that.

I'll squashed the last commit before merging. :)

File: pkg/kubelet/reason_cache.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+// maxReasonCacheEntries is the cache entry number in lru cache. 1000 is a proper number
+// for our 100 pods per node target. If we support more pods per node in the future, we
+// may want to increase the number.
+const maxReasonCacheEntries = 1000
+
+func NewReasonCache() *ReasonCache {
+	return &ReasonCache{cache: lru.New(maxReasonCacheEntries)}
+}
+
+func (c *ReasonCache) getKey(uid types.UID, name string) string {
```
Comment (yifan-gu) [inline pkg/kubelet/reason_cache.go]: getKey sounds ambiguous , cacheKey or something?

Comment (Random-Liu) [inline pkg/kubelet/reason_cache.go]: ACK. I'll use the function name in code version reason cache `composeKey`.

**Comment by Random-Liu** [comment]
@yujuhong This PR is ready for review and could be merged in parallel with pod cache. :)

**Comment by yujuhong** [comment]
> @yujuhong The bug should have been fixed by the new commit.
> 
> This PR broke e2e test "should have their auto-restart back-off timer reset on image update" at first.
> 
> After some debuging, I found that we should call reasonCache.Update() whenever there is error or not. Because even when there is no error, we should "update" the reason cache to clean up the old error reason.
> 
> If we don't do that, once a container is failed to be started, its error reason will never be cleared, and will cause convertStatusToAPIStatus() always return Waiting state after that.
> 
> I'll squashed the last commit before merging. :)

Yes, the reason cache needs to be updated always. Sorry I didn't catch that in the previous review.

**Comment by Random-Liu** [comment]
@yujuhong No, it's my fault. :)

**Comment by Random-Liu** [comment]
@k8s-bot test this please.

**Comment by Random-Liu** [comment]
Rebased, will apply LGTM again after all the test passed.

**Pull request merged**